File: ekosis/state_keepers/error_state_list.py
import json

from typing import Dict, List, Any

from .error_state_keeper import ErrorStateKeeper

from ..util.singleton import SingletonType

# --------------------------------------------------------------------------------
class ErrorStateList(metaclass=SingletonType):
    error_state_map: Dict[str, ErrorStateKeeper] = {}

    def __init__(self):
        pass

    def __str__(self) -> str:
        return self.to_json()

    def has_error_id(self, error_id: str) -> bool:
        error_id_upper = error_id.upper()
        return error_id_upper in self.error_state_map.keys()

    def to_json(self) -> str:
        error_list   : List[ErrorStateKeeper] = self.get_error_states()
        response_list: List[Dict[str, Any]]   = []
        for error in error_list:
            response_list.append(error.to_dict())
        return json.dumps(response_list)

    def increment(self, error_id: str, description: str):
        error_id_upper = error_id.upper()
        if error_id_upper not in self.error_state_map.keys():
            self.error_state_map[error_id_upper] = ErrorStateKeeper(error_id_upper, description)

        self.error_state_map[error_id_upper].increment()

    def clear_some(self, error_id: str, how_many: int):
        error_id_upper = error_id.upper()
        if error_id_upper in self.error_state_map.keys():
            self.error_state_map[error_id_upper].clear_some(how_many)

    def clear_all(self, error_id: str):
        error_id_upper = error_id.upper()
        if error_id_upper in self.error_state_map.keys():
            self.error_state_map[error_id_upper].clear_all()

    def get_error_states(self) -> List[ErrorStateKeeper]:
        result: List[ErrorStateKeeper] = []
        for error_state in self.error_state_map.values():
            if error_state.is_set():
                result.append(error_state)
        return result

File: ekocpp/src/state_keepers/error_state_list.cpp
#include "error_state_list.hpp"

#include <algorithm>

namespace {

std::string to_upper(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c) { return std::toupper(c); });
    return s;
}

} // namespace

ErrorStateList& ErrorStateList::instance() {
    static ErrorStateList list;
    return list;
}

bool ErrorStateList::has_error_id(const std::string& error_id) const {
    return error_state_map_.find(to_upper(error_id)) != error_state_map_.end();
}

void ErrorStateList::increment(const std::string& error_id, const std::string& description) {
    const std::string error_id_upper = to_upper(error_id);
    auto it = error_state_map_.find(error_id_upper);
    if (it == error_state_map_.end()) {
        it = error_state_map_.emplace(error_id_upper, ErrorStateKeeper(error_id_upper, description)).first;
    }
    it->second.increment();
}

void ErrorStateList::clear_some(const std::string& error_id, int how_many) {
    auto it = error_state_map_.find(to_upper(error_id));
    if (it != error_state_map_.end()) {
        it->second.clear_some(how_many);
    }
}

void ErrorStateList::clear_all(const std::string& error_id) {
    auto it = error_state_map_.find(to_upper(error_id));
    if (it != error_state_map_.end()) {
        it->second.clear_all();
    }
}

std::vector<nlohmann::json> ErrorStateList::get_error_states_json() const {
    std::vector<nlohmann::json> result;
    for (const auto& [key, error_state] : error_state_map_) {
        if (error_state.is_set()) {
            result.push_back(error_state.to_json());
        }
    }
    return result;
}
